daemon/graphdriver/copy/copy.go

```diff
@@ -107,156 +120,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
-				if err2 := os.Link(srcPath, dstPath); err2 != nil {
-					return err2
+	// Just a dumb wrapper around the above function which does the heavy lifting
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case fci, ok := <-fileCopyInfoChan:
+			if !ok {
+				return nil
+			}
+			switch fci.mode {
+			case dirMode:
+				if err := copyMetadataShared(fci.dstPath, fci.srcPath, fci.fileInfo, fci.stat, copyXattrs); err != nil {
+					return err
 				}
-			} else if hardLinkDstPath, ok := copiedFiles[id]; ok {
-				if err2 := os.Link(hardLinkDstPath, dstPath); err2 != nil {
-					return err2
+			case fileMode:
+				if err := doFileCopy(fci.srcPath, fci.dstPath, fci.fileInfo, fci.stat, fci.sharedInode, &copyWithFileRange, &copyWithFileClone, copyXattrs, copyMode); err != nil {
+					return err
 				}
-			} else {
-				if err2 := copyRegular(srcPath, dstPath, f, &copyWithFileRange, &copyWithFileClone); err2 != nil {
-					return err2
+			case updateDirTimeMode:
+				ts := []syscall.Timespec{fci.stat.Atim, fci.stat.Mtim}
+				if err := system.LUtimesNano(fci.dstPath, ts); err != nil {
+					return err
 				}
-				copiedFiles[id] = dstPath
 			}
+		}
 
-		case os.ModeDir:
-			if err := os.Mkdir(dstPath, f.Mode()); err != nil && !os.IsExist(err) {
-				return err
-			}
+	}
+}
 
-		case os.ModeSymlink:
-			link, err := os.Readlink(srcPath)
-			if err != nil {
-				return err
-			}
+// readDirNames reads the directory named by dirname and returns
+// a sorted list of directory entries.
+func readDirNames(dirname string) ([]string, error) {
+	f, err := os.Open(dirname)
+	if err != nil {
+		return nil, err
+	}
+	names, err := f.Readdirnames(-1)
+	f.Close()
+	if err != nil {
+		return nil, err
+	}
+	return names, nil
+}
 
-			if err := os.Symlink(link, dstPath); err != nil {
-				return err
-			}
+type walkFn struct {
+	ctx            context.Context
+	srcDir         string
+	dstDir         string
+	workers        []chan *fileCopyInfo
+	rand           *rand.Rand
+	sharedInodeMap map[fileID]*sharedInode
+}
 
-		case os.ModeNamedPipe:
-			fallthrough
-		case os.ModeSocket:
-			if err := unix.Mkfifo(dstPath, stat.Mode); err != nil {
-				return err
-			}
+func (w *walkFn) rebasePath(srcPath string) (string, error) {
+	// Rebase path
+	relPath, err := filepath.Rel(w.srcDir, srcPath)
+	if err != nil {
+		return "", err
+	}
 
-		case os.ModeDevice:
-			if rsystem.RunningInUserNS() {
-				// cannot create a device if running in user namespace
-				return nil
-			}
-			if err := unix.Mknod(dstPath, stat.Mode, int(stat.Rdev)); err != nil {
-				return err
-			}
+	return filepath.Join(w.dstDir, relPath), nil
+}
 
-		default:
-			return fmt.Errorf("unknown file type for %s", srcPath)
-		}
+func (w *walkFn) walkFunc(srcPath string, f os.FileInfo, c chan *fileCopyInfo, err error) error {
+	var sInode *sharedInode
+	mode := fileMode
+	if err != nil {
+		return err
+	}
+	// Check if the errgroup has gotten canceled / shutdown
+	err = w.ctx.Err()
```

Comment:
This is just there to keep the code as "clean" as possible, and to make the code as similar to the code in the stdlib. It could get passed in as well.

Reply:
We should pass in in general. I think it's fine to deviate from the stdblib interface since we do have a context to pass around.